### PR TITLE
Add Datadog metrics exporter

### PR DESCRIPTION
**Pull request by mx-psi**
**Description:** 

- Add [Datadog](https://www.datadoghq.com/) metrics exporter using [DogStatsD](https://docs.datadoghq.com/developers/metrics/dogstatsd_metrics_submission) via the [`statsd`](https://pkg.go.dev/github.com/DataDog/datadog-go/statsd?tab=doc) package.

**Link to tracking Issue:** n/a

**Testing:** 

- Added unit tests.
- Built the Collector on this branch and performed manual tests with different receivers.

**Documentation:**

- The README briefly documents what the exporter does.
- The example configuration file documents every option possible values and usage.

### CODE REVIEW

**Comment by mx-psi** [comment]
This can be reviewed now: Codecov patch is failing but I think the patch coverage is reasonable (88.38%); some of the lines that Codecov indicates can't be tested (since we are not doing any configuration sanitization right now).

File: exporter/datadogexporter/factory.go
Comment (keitwb) [inline exporter/datadogexporter/factory.go:95]: You can omit this function and the corresponding WithTraces option above and it will by default consider traces unsupported.

Comment (mx-psi) [reply exporter/datadogexporter/factory.go]: Makes sense, I removed it

File: exporter/datadogexporter/config.go
Comment (keitwb) [inline exporter/datadogexporter/config.go:31]: Is this like internal metrics about statsd or all metrics in general?  I would clarify the description of this.

**Review by keitwb — APPROVED** [review]
LGTM

**Comment by keitwb** [comment]
@mx-psi do try and improve coverage a bit.

**Comment by albertvaka** [comment]
👋 Who else should review this before it can be merged? @asuresh4 (since you are assigned)?

**Comment by asuresh4** [comment]
> 👋 Who else should review this before it can be merged? @asuresh4 (since you are assigned)?

Did a quick pass through the PR on Friday and I think it looks good. I plan on taking a closer look today.

**Review by asuresh4 — APPROVED** [review]

**Comment by bogdandrutu** [comment]
@tylerbenson as one of the DataDog person can you review this? Or anyone else from DataDog can do it?

File: exporter/datadogexporter/factory.go
Comment (bogdandrutu) [inline exporter/datadogexporter/factory.go:53]: This does not need to be public

**Comment by tylerbenson** [comment]
I'll ask around to find someone that can review.

**Comment by ericmustin** [comment]
@tylerbenson @mx-psi @bogdandrutu @albertvaka 

Hey friends. Just chiming in here, @mx-psi @tylerbenson and @albertvaka are all colleagues of mine at Datadog. I'm an approver on [opentelemetry-ruby](https://github.com/open-telemetry/opentelemetry-ruby) but am quite new to the collector, so I don't think I can add much to the code review, however generally I think this looks good to me.  I know that some of the metrics folks have been reviewing things over on [our fork](https://github.com/DataDog/opentelemetry-collector-contrib) and so I have a reasonable level of confidence here that Datadog as an org is both comfortable with this exporter and also plans to support it and add to it over time, should any issues arise. Additionally, I know there are a few limitations currently between what dogstatsd allows us to do and what our metric api / otel metrics permit, so while this PR might not have full coverage for every metric type,  I think Datadog plans to add additional (ie agentless) export options in the future, along with (ideally) exporters for traces at the collector level. I'll try to pop in on the SIG mtg later to say hello as I have a few other unrelated questions.

 I hope that helps! certainly please think of @mx-psi as the voice of reason here for all things datadog metrics going forward

File: exporter/datadogexporter/config.go
Comment (bogdandrutu) [inline exporter/datadogexporter/config.go:28]: Can you use the https://github.com/open-telemetry/opentelemetry-collector/blob/master/config/confignet/confignet.go#L44 or https://github.com/open-telemetry/opentelemetry-collector/blob/master/config/confignet/confignet.go#L22?

**Review by bogdandrutu — COMMENTED** [review]
There are some reviews not addressed. Can you also split this based on https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#when-adding-a-new-component

**Comment by mx-psi** [comment]
@bogdandrutu 

> There are some reviews not addressed. Can you also split this based on https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#when-adding-a-new-component

Sure! I will address the remaining comments next week and split it up into different PRs following the recently updated contributing guidelines. We plan on expanding the exporter's functionality so the configuration will not be final.

Is there any other recent change I should take into account for next week? I have tried to keep everything up to date with the rest of the exporters, but it's a bit hard to keep up with all the changes going on in the Collector.

> @/tylerbenson as one of the DataDog person can you review this? Or anyone else from DataDog can do it? 

As Eric stated in his message, this has already been reviewed by Datadog folks over at our fork; I will ask internally for people to chime in into the separate PRs next week.

File: exporter/datadogexporter/config.go
Comment (mx-psi) [reply exporter/datadogexporter/config.go:28]: The `statsd` library picks up which protocol to use automatically (`udp` or `unixgram`) based on the `endpoint` field shape, so the `transport` field would have to be ignored in some cases (in particular, if `endpoint` starts with `unix://` it's considered UDS, and if certain environment variables are set it would be UDP). Would that behavior be okay you? Otherwise we would have to leave it as it is right now.

**Comment by mx-psi** [comment]
Hi all, after some internal discussion we have decided to rework this PR a bit so I am closing this. I will open new PRs in the future following the contribution guidelines with the reworked configuration and implementation. Thanks for the reviews, I will take them into account for the future PRs!

**Pull request closed without merging**
